### PR TITLE
Feat/ssset better deletion

### PR DESCRIPTION
**Pull request by cristiGuranIonos**
### Description of your changes

This PR adds:

## Checklist

I have:

- [ ] Add PR name as appropriate (e.g. `feat`/`fix`/`doc`/`test`/`refactor`)
- [ ] Run `make reviewable` and `make crds.clean` to ensure the PR is ready for review
- [ ] Add or update tests (if applicable)
- [ ] Add or update Documentation using `make docs.update` (if applicable)
- [ ] Update `docs/CHANGELOG.md` file (label: `upcoming release`)
- [ ] Check Sonar Cloud Scan
- [ ] Update Github or Jira Issue

### CODE REVIEW

File: internal/controller/compute/statefulserverset/statefulserverset.go [hunk 1]
```diff
@@ -293,31 +293,35 @@ func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
 	}
 	cr.SetConditions(xpv1.Deleting())
 
-	e.log.Info("Deleting the DataVolumes with label", "label", cr.Name)
-	if err := e.kube.DeleteAllOf(ctx, &v1alpha1.Volume{}, client.InNamespace(cr.Namespace), client.MatchingLabels{
-		statefulServerSetLabel: cr.Name,
-	}); err != nil {
-		return err
+	// e.dataVolumeController.Delete()
```
Comment (cbeti-ionos) [inline internal/controller/compute/statefulserverset/statefulserverset.go:296]: will this encompass the code below?

File: internal/controller/serverset/serverset.go [hunk 2]
```diff
@@ -192,16 +199,70 @@ func (e *external) populateReplicasStatuses(ctx context.Context, cr *v1alpha1.Se
...
+			// 			globalStateMap[cr.Name] = substitution.GlobalState{}
+			// 		}
+			// 		if !globalStateMap[cr.Name].Exists(identifier, subst.Key) {
+			// 			globalStateMap[cr.Name].Set(identifier, subst.Key, value)
+			// 			e.log.Info("substitution value updated in global state", "serverset name", cr.Name, "for key", subst.Key, "and value", value)
+			// 		}
+			// 	} else {
+			// 		e.log.Info("substitution value not found", "serverset name", cr.Name, "for key", subst.Key)
+			// 	}
+			// }
```
Comment (cbeti-ionos) [inline internal/controller/serverset/serverset.go:231]: is this still needed ?

File: internal/controller/serverset/serverset.go [hunk 3]
```diff
@@ -484,31 +545,31 @@ func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
...
+			return err
+		}
+		if err := e.serverController.Delete(ctx, getNameFrom(cr.Spec.ForProvider.Template.Metadata.Name, replicaIndex, serverVersion), cr.Namespace); err != nil {
+			return err
+		}
+		for nicIndex := range cr.Spec.ForProvider.Template.Spec.NICs {
+			if err := e.nicController.Delete(ctx, getNicName(cr.Spec.ForProvider.Template.Spec.NICs[nicIndex].Name, replicaIndex, nicIndex, serverVersion), cr.Namespace); err != nil {
+				return err
+			}
+		}
```
Comment (cbeti-ionos) [inline internal/controller/serverset/serverset.go:565]: previously, the nic deletion happened before server deletion, does the order matter ?

File: internal/controller/serverset/serverset_test.go [hunk 4]
```diff
@@ -1349,116 +1349,113 @@ func Test_serverSetController_updateOrRecreateVolumes_activeReplicaUpdatedLast_c
 	assertions.Equal("nic1-0-0-0", nicController.lastMethodCall[deleteMethod][secondArg])
 }
 
-func Test_serverSetController_Delete(t *testing.T) {
```
Comment (cbeti-ionos) [inline internal/controller/serverset/serverset_test.go:1352]: is this test temporarily removed ?

File: internal/controller/serverset/serverset.go (hunk 3 above)
Comment (cristiGuranIonos) [reply internal/controller/serverset/serverset.go:565]: yes, in some cases deleting before threw an error

File: internal/controller/serverset/serverset_test.go (hunk 4 above)
Comment (cristiGuranIonos) [reply internal/controller/serverset/serverset_test.go:1352]: Probably permanently. It wasn't really testing much

**Review by cbeti-ionos — APPROVED** [review]

**Pull request merged**